Libraries/Opc.Ua.Configuration/IApplicationInstanceBuilder.cs

```diff
@@ -0,0 +1,167 @@
+    /// </summary>
+    public interface IApplicationInstanceBuilder :
+        IApplicationInstanceBuilderTypes,
+        IApplicationInstanceBuilderServerSelected,
+        IApplicationInstanceBuilderClientSelected,
+        IApplicationInstanceBuilderSecurity,
+        IApplicationInstanceBuilderServerPolicies,
+        IApplicationInstanceBuilderCreate
+    {
+    };
+
+    /// <summary>
+    /// The server configuration.
+    /// </summary>
+    public interface IApplicationInstanceBuilderTypes :
+        IApplicationInstanceBuilderServer,
+        IApplicationInstanceBuilderClient
+    {
+    }
+
+    /// <summary>
+    /// The server configuration.
+    /// </summary>
+    public interface IApplicationInstanceBuilderServerSelected :
+        IApplicationInstanceBuilderServerPolicies,
+        IApplicationInstanceBuilderClient,
+        IApplicationInstanceBuilderSecurity
+    {
+    }
+
+    /// <summary>
+    /// The server configuration.
+    /// </summary>
+    public interface IApplicationInstanceBuilderClientSelected :
+        IApplicationInstanceBuilderSecurity
+    {
+    }
+
+    /// <summary>
+    /// The server configuration.
+    /// </summary>
+    public interface IApplicationInstanceBuilderServer
+    {
+        /// <summary>
+        /// Configure instance to be used for UA server.
+        /// </summary>
+        IApplicationInstanceBuilderServerSelected AsServer(
+            string[] baseAddresses,
+            string[] alternateBaseAddresses = null);
+    }
+
+    /// <summary>
+    /// The server configuration.
+    /// </summary>
+    public interface IApplicationInstanceBuilderClient
+    {
+        /// <summary>
+        /// Configure instance to be used for UA client.
+        /// </summary>
+        IApplicationInstanceBuilderClientSelected AsClient();
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    public interface IApplicationInstanceBuilderSecurity
+    {
+        /// <summary>
+        /// Add the security configuration.
+        /// </summary>
+        /// <remarks>
+        /// The pki root path default to the directory store
+        /// location as defined in <see cref="CertificateStoreIdentifier.DefaultPKIRoot"/>
+        /// A <see cref="CertificateStoreType"/> defaults to the corresponding default store location.
+        /// </remarks>
+        /// <param name="subjectName">Application certificate subject name as distinguished name. A DC=localhost entry is converted to the hostname. The common name CN= is mandatory.</param>
+        /// <param name="pkiRoot">The path to the pki root.</param>
+        /// <param name="appRoot">The path to the app cert store, if different than the pki root.</param>
+        /// <param name="rejectedRoot">The path to the rejected certificate store.</param>
+        IApplicationInstanceBuilderCreate AddSecurityConfiguration(
+            string subjectName,
+            string pkiRoot = null,
```

Comment:
What will be used if the paths (optional arguments) are set to null?
- directory in user folder?
- directory relative to executable?
- system default?
- ...